### PR TITLE
Adding IWSLT 2017  dataset.

### PR DESCRIPTION
**Pull request by Narsil**
Created a [IWSLT 2017](https://sites.google.com/site/iwsltevaluation2017/TED-tasks) dataset script for the *multilingual data*.

```
Bilingual data:  {Arabic, German, French, Japanese, Korean, Chinese} <-> English
Multilingual data:   German, English, Italian, Dutch, Romanian. (Any pair)
```

I'm unsure how to handle bilingual vs multilingual. Given `nlp` architecture a Config option seems to be the way to go, however, it might be a bit confusing to have different language pairs with different option. Using just language pairs is not viable as English to German exists in both.

Any opinion on how that should be done ?
EDIT: I decided to just omit de-en from multilingual as it's only a subset of the bilingual one. That way only language pairs exist.
EDIT : Could be interesting for #438

### CODE REVIEW

File: datasets/iwslt2017.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+
+    def _split_generators(self, dl_manager):
+        """Returns SplitGenerators."""
+        source, target = self.config.pair.split("-")
+        if self.config.is_multilingual:
+            dl_dir = dl_manager.download_and_extract(MULTI_URL)
+            data_dir = os.path.join(dl_dir, "DeEnItNlRo-DeEnItNlRo")
+            years = [2010]
+        else:
+            bi_url = f"https://wit3.fbk.eu/archive/2017-01-trnted//texts/{source}/{target}/{source}-{target}.tgz"
```
Comment (lhoestq) [inline datasets/iwslt2017.py:125]:
```suggestion
            bi_url = f"https://wit3.fbk.eu/archive/2017-01-trnted/texts/{source}/{target}/{source}-{target}.tgz"
```

File: datasets/iwslt2017.py [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+    if source != target
+]
+BI_PAIRS = [
+    f"{source}-{target}"
+    for source in BI_LANGUAGES
+    for target in BI_LANGUAGES
+    if source != target and (source == "en" or target == "en")
+]
+
+PAIRS = MULTI_PAIRS + BI_PAIRS
```
Comment (lhoestq) [inline datasets/iwslt2017.py:78]: I think it's fine this way

File: datasets/iwslt2017.py [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+    ]
+
+    def _info(self):
+        return nlp.DatasetInfo(
+            # This is the description that will appear on the datasets page.
+            description=_DESCRIPTION,
+            # nlp.features.FeatureConnectors
+            features=nlp.Features(
+                {"source": nlp.Value("string"), "target": nlp.Value("string"),}
+            ),
```
Comment (lhoestq) [inline datasets/iwslt2017.py:107]: I would use the `Translation` feature type.

You can see an example of usage in the [FLoRes dataset script](https://github.com/huggingface/nlp/blob/master/datasets/flores/flores.py#L93)

File: datasets/iwslt2017.py [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+                                source_row = source_row.strip()
+                                target_row = target_row.strip()
+                            else:
+                                continue
+
+                        yield id_, {
+                            "source": source_row,
+                            "target": target_row,
+                        }
```
Comment (lhoestq) [inline datasets/iwslt2017.py:226]: If you use the `Translation` feature type defined this way

```python
features=nlp.Features({"translation": nlp.features.Translation(languages=self.config.pair.split("-"))})
```

Then the `yield` statement here would become:

```python
yield id_, {
    "translation": {
        source: source_row,
        target: target_row
    }
}
```

where source and target are defined by
```python
source, target = self.config.pair.split("-")
```

Comment (Narsil) [reply datasets/iwslt2017.py:226]: Ok done.

**Comment by Narsil** [comment]
Ok I tried to add the dummy dataset (I actually modified the dummy_data command to generate them for me because it was too painful to do that manually).

The dummy_data test seems to work:
```bash
RUN_SLOW=1 pytest tests/test_dataset_common.py::LocalDatasetTest::test_load_dataset_all_configs_iwslt2017
```

However the test on the full data fails, because the `**config_kwargs` don't include `pair, multilingual`.
I could add a default parameter for the Config (but that feels broken, how can one config be the "default" ?). If I do I still have errors, saying that something within the downloader is a directory so I'm not sure where that comes from.

I can share my auto_zip dummy data code if you want (I tried to keep it clean). The way it works is that it just keeps X line from the beginning of the original files, and Y lines at the end. It's good enough for my usage, but I guess it could work for most data files out there (as long as they're real text and not binary format)

**Comment by lhoestq** [comment]
The slow test doesn't support dataset that require config parameters that don't have default values.

To improve that we can replace it by two tests:
- one test that loads the default config (it can simply be the first config of the config lists for example)
- one tests that iterate over all configs and load them all one by one

By using the configs inside the builder config lists, there is no need to instantiate new configs, so the missing parameter error doesn't happen.

Does that sound good to you ?

**Comment by Narsil** [comment]
Seems fair.
However I'm unsure what I should do ?

Should I wait for #527 to pass and rebase and the command will be the same ?
Should I update something ?

**Comment by lhoestq** [comment]
I think everything is fine on your side. Thanks for adding this dataset :)

I think it's better to wait for the slow test to be updated if you don't mind.

**Comment by Narsil** [comment]
Sure ! :)

File: datasets/iwslt2017/iwslt2017.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+        """
+
+        Args:
+            pair: the language pair to consider
+            is_multilingual: Is this pair in the multilingual dataset (download source is different)
+            **kwargs: keyword arguments forwarded to super.
+        """
+        self.pair = pair
+        self.is_multilingual = is_multilingual
+        super(IWSLT2017Config, self).__init__(**kwargs)
```
Comment (julien-c) [inline datasets/iwslt2017/iwslt2017.py:60]:
```suggestion
        super().__init__(**kwargs)
```

We're Python 3 only so you can omit this

File: datasets/iwslt2017/iwslt2017.py [hunk 6]
```diff
@@ -166,8 +166,8 @@ def _generate_examples(self, source_files, target_files, split):
         id_ = 0
         source, target = self.config.pair.split("-")
         for source_file, target_file in zip(source_files, target_files):
-            with open(source_file) as sf:
-                with open(target_file) as tf:
+            with open(source_file, "r", encoding="utf-8") as sf:
+                with open(target_file, "w", encoding="utf-8") as tf:
```
Comment (julien-c) [inline datasets/iwslt2017/iwslt2017.py:170]:
```suggestion
                with open(target_file, "r", encoding="utf-8") as tf:
```

Comment (Narsil) [reply datasets/iwslt2017/iwslt2017.py:170]: Bitten by my own internal pattern matching (target files get written to)

File: src/nlp/arrow_writer.py [hunk 7]
```diff
@@ -362,6 +362,7 @@ def finalize(self, metrics_query_result: dict):
                 that the filter keeps only the metrics for the considered split, under the namespace `split_name`.
         """
         import apache_beam as beam
+
```
Comment (lhoestq) [inline src/nlp/arrow_writer.py:365]: It looks like an isort issue

File: src/nlp/builder.py [hunk 8]
```diff
@@ -958,6 +958,7 @@ def _build_pcollection(pipeline, extracted_dir):
     def _download_and_prepare(self, dl_manager, verify_infos):
         # Create the Beam pipeline and forward it to _prepare_split
         import apache_beam as beam
+
```
Comment (lhoestq) [inline src/nlp/builder.py:961]: same here

File: src/nlp/search.py [hunk 9]
```diff
@@ -14,8 +14,8 @@
 
 try:
     import elasticsearch as es
-    from elasticsearch import Elasticsearch
     import elasticsearch.helpers
+    from elasticsearch import Elasticsearch
```
Comment (lhoestq) [inline src/nlp/search.py:18]: same here

**Review by lhoestq — APPROVED** [review]
Looks good :)
Could you remove the changes caused by `isort` ? (see my comments)
We haven't updated it yet

File: src/nlp/arrow_writer.py (hunk 7 above)
Comment (Narsil) [reply src/nlp/arrow_writer.py]: Okay, I removed them (automated tools are eager to do stuff sometimes :))

**Comment by lhoestq** [comment]
Thanks for fixing the isort/black changes :)
Feel free to merge if it's good for you @Narsil

**Pull request merged**